azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,348 @@
+
+    def _remove_extensions_if_necessary(self, extensions_config):
+        """
+        If this is a Fabric GS, but the InSvdSeqNo did NOT change, then the goal state was
+        created directly by Fabric and bypassed CRP. A common scenario is remote access.
+        Another is when wire server restarts and uses a new incarnation.
+        The problem is any extensions contained here may be out of date, because they were
+        more recently updated via a FastTrack GS. Therefore, we remove them here in that case.
+        """
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            svd_seqNo = self._get_svd_seqNo()
+            if str(extensions_config.svd_seqNo) == str(svd_seqNo):
+                logger.info("SvdSeqNo did not change. Removing extensions from goal state")
+                extensions_config.ext_handlers = None
+                self._last_svd_seqNo = extensions_config.svd_seqNo
+                self._pending_svd_seqNo = extensions_config.svd_seqNo
+            else:
+                self._pending_svd_seqNo = svd_seqNo
+
+    def _decide_what_to_process(self, fabric_changed, fast_track_changed):
+        """
+        If just Fabric GS changed, then process only that.
+        If just FastTrack GS changed, then process only that.
+        If both changed, then process Fabric and then FastTrack.
+        If neither changed, then process whichever we used last (to keep with the current behavior)
+        """
+        if fabric_changed:
+            self._set_reason("FabricChanged")
+            return GOAL_STATE_SOURCE_FABRIC
+        if fast_track_changed:
+            self._set_reason("FTChanged")
+            return GOAL_STATE_SOURCE_FASTTRACK
+        if self._get_mode() == GOAL_STATE_SOURCE_FASTTRACK:
+            self._set_reason("LastFT")
+            return GOAL_STATE_SOURCE_FASTTRACK
+
+        self._set_reason("LastFabric")
+        return GOAL_STATE_SOURCE_FABRIC
+
+    def _set_reason(self, reason):
+        self._reason = "{0}: FastTrack={1}, Fabric={2}".format(reason, self._ft_changed_detail, self._fabric_changed_detail)
+
+    def _get_fast_track_changed(self, artifacts_profile):
+        if artifacts_profile is None:
+            self._ft_changed_detail = "NoProfile"
+            return False
+        if not artifacts_profile.has_extensions():
+            self._ft_changed_detail = "NoExtensions"
+            return False
+
+        sequence_number = self._last_seqNo
+        if sequence_number is None:
+            sequence_number = self._get_sequence_number()
+        if sequence_number is None or sequence_number != artifacts_profile.get_sequence_number():
+            self._ft_changed_detail = "seqNoChanged"
+            return True
+
+        self._ft_changed_detail = "NoChange"
+        return False
+
+    def _get_fabric_changed(self, goal_state_incarnation):
+        if goal_state_incarnation is None:
+            self._fabric_changed_detail = "NoInc"
```

Comment:
An integer would be problematic here because it is less readable and would make searching the code more difficult. Ultimately we don't have any logic that looks at these values (and then at least a constant would make sense), but they're written to the logs and designed for debuggability.

Reply:
Sorry bad example from my end, I mainly meant something like - 
```python
class Reasons:
    FT_CHANGE_NO_PROFILE = "NoProfile"
    FT_CHANGE_SEQ_NO_CHANGED = "seqNoChanged"
```

Something like that. I agree its a low hanging fruit, good to have but definitely not must have.

Reply:
ok. changed